### PR TITLE
`AppSideNav` - Follow-on improvements for headers & aria (HDS-3897)

### PR DESCRIPTION
**Pull request by KristinLBradley**
### :pushpin: Summary

If merged, this PR will make improvements to the `AppSideNav` component related to `List::Title` headers and the `ToggleButton` aria properties. 

(It will merge into the component branch, not directly into the main branch.)

### :hammer_and_wrench: Detailed description

**Implemented changes:**

* Change `List::Title` to h3 & add visually hidden h2 to AppSideNav - [Reference comment](https://github.com/hashicorp/design-system/pull/2384#discussion_r1764009486)
* Improve ariaLabel for `ToggleButton` state & use aria-expanded - [Reference comment](https://github.com/hashicorp/design-system/pull/2384#discussion_r1764017267)

### :link: External links

* Jira ticket: [HDS-3897](https://hashicorp.atlassian.net/browse/HDS-3897)

***

### 👀 Component checklist

- [x] Percy was checked for any visual regression
- ~~[ ] A changelog entry was added via [Changesets]~~(https://github.com/changesets/changesets) if needed (see [templates here](https://hashicorp.atlassian.net/wiki/spaces/HDS/pages/3243114706/Changelog+authoring+best+practices#Templates))

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

[HDS-3897]: https://hashicorp.atlassian.net/browse/HDS-3897?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/components/src/components/hds/app-side-nav/base.hbs
Comment (KristinLBradley) [inline packages/components/src/components/hds/app-side-nav/base.hbs:7]: Let me know if this header should use different wording, be moved elsewhere in the DOM, and/or have the text exposed in case consumers would want to override it, etc.

Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/base.hbs:7]: @shleewhite @MelSumner

File: packages/components/src/components/hds/app-side-nav/list/title.hbs
Comment (KristinLBradley) [inline packages/components/src/components/hds/app-side-nav/list/title.hbs:10]: I change the div tags to h3 tags for now.

Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs:10]: I'm not sure how I would be able to associate the ul with the title using aria-labelled by with the current structure. I think the whole structure would have to be redone to make it work. 

(I worked on the SideNav in the early stages and wanted to structure it as headers outside of uls which served as the title of the uls but was overruled in favor of the current structure. The structure has complexities due to some of how it functions within Products I believe.)

File: packages/components/src/components/hds/app-side-nav/index.hbs
Comment (KristinLBradley) [inline packages/components/src/components/hds/app-side-nav/index.hbs:22]: Let me know if this label is ok or if it should be changed, moved, and/or exposed so consumers can override

File: packages/components/src/components/hds/app-side-nav/list/title.hbs
Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs:10]: If we really need to do this (aria-labelled changes to associate the Title headers with the Lists) then I could make another branch of the component PR as it's going to require more extensive changes and testing than the other smaller updates I've made in this current branch. Please let me know what you think.

File: packages/components/src/components/hds/app-side-nav/base.hbs
Comment (MelSumner) [reply packages/components/src/components/hds/app-side-nav/base.hbs:7]: Should this be "Application navigation" ? 

Also, is there an aria-label on the nav element in the sidebar? It might be redundant to have both; I'd update in these ways:

1. Add an `id` to the `h2`
2. change the `aria-label` in the aside nav to be `aria-labelledby` and give it the value of the `h2`'s `id` attribute.

File: packages/components/src/components/hds/app-side-nav/list/title.hbs
Comment (MelSumner) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs:10]: 1. The lists do need to have accNames or else they just show up as "list" in the rotor
2. I think it would be sufficient to have this list item given an `id` attribute and then use `aria-labelledby` on the `ul`. 

I don't think these should be h3's.

File: packages/components/src/components/hds/app-side-nav/base.hbs
Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/base.hbs:7]: The `nav` element which wraps each list doesn't have an aria-label (I could add an aria-labelledby though/). There can be multiple `nav` elements potentially and likely. The `appFrame` `sidebar` child component where the `AppSideNav` slots into doesn't have an aria-label.

Does it matter if the h2 labels multiple nav elements?

Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/base.hbs]: I updated the h2 text to "Application local navigation" since the AppSideNav is supposed to contain page level navigation and to be paired with the AppHeader which contains "global" & "utility" navigation. Let me know if I should change the text to something else again.

I added a hard-coded id to the h2 and aria-labelledby with the id valu to the `nav` elements wrapping lists of links. Again there are likely to be multiple nav elements. (I hard-coded the ids instead of generating since it didn't seem like we would really need to generate them but I can change that too if desired.)

Please let me know if these changes are ok or if I should make more. @MelSumner @shleewhite @alex-ju

File: packages/components/src/components/hds/app-side-nav/list/title.hbs
Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs:10]: Can you take a look at the code and let me know if you have ideas on how to do it with the current structure? Redoing the current structure if something I really want to avoid unless absolutely necessary.

If the current structure is unacceptable on the AppSideNav it will need to be updated on the SideNav too.

@shleewhite @alex-ju Please let me know if you have any input on this.

File: packages/components/src/components/hds/app-side-nav/index.hbs
Comment (shleewhite) [reply packages/components/src/components/hds/app-side-nav/index.hbs:22]: Hmm.. I think it should match what the h2 calls it for consistency - so "Application local navigation"

File: packages/components/src/components/hds/app-side-nav/list/title.hbs
Comment (shleewhite) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs:10]: Hmmm.. I hadn't realized before that it is just one list with everything inside it... this is tough because visually theyre either separate lists or sub-lists, both of which would require refactoring parts of the structure. 

Can you clarify why you would need to update it in sidenav too? Isn't it going to be deprecated?

File: packages/components/src/components/hds/app-side-nav/base.hbs
Comment (shleewhite) [reply packages/components/src/components/hds/app-side-nav/base.hbs]: I think its fine.... does the AppHeader navigation have an accessible name?

File: packages/components/src/components/hds/app-side-nav/index.hbs
Comment (alex-ju) [reply packages/components/src/components/hds/app-side-nav/index.hbs:22]: in this case, I suggest we point to the heading to avoid duplicating content
```suggestion
        aria-labelledby="hds-app-side-nav-header"
```

**Review by alex-ju — COMMENTED** [review]
This looks good overall! I will wait to see if we come up with a pragmatic solution for [titles in the list](https://github.com/hashicorp/design-system/pull/2463/files#r1779208065). Otherwise we'll record the need for improvements in a follow-up tasks.

File: packages/components/src/components/hds/app-side-nav/list/title.hbs
Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs:10]: Update: From squad discussions, we will avoid making new updates the the SideNav unless absolutely necessary.

File: packages/components/src/components/hds/app-side-nav/base.hbs
Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/base.hbs]: If the AppHeader needs to have a header or aria-label or similar added could you please open a new issue or PR? I want to focus on the AppSideNav for the current open PRs.

File: packages/components/src/components/hds/app-side-nav/list/title.hbs
Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs:10]: Let me know if you have ideas on improvements that could be made to the current structure or if the updates I already made are ok for now.

**Comment by KristinLBradley** [comment]
I'm getting this error now since there are multiple AppSideNav components on the Showcase page although in actual usage there would only ever be one per page:

```
Error: The page should have no accessibility violations. Violations:
            [moderate]: Ensures landmarks are unique 
            Violated 1 time. Offending nodes are: 
            <nav class="hds-app-side-nav__list-wrapper" aria-labelledby="hds-app-side-nav-header" aria-label="Primary on portal demo 1">
```

**Comment by alex-ju** [comment]
> I'm getting this error now since there are multiple AppSideNav components on the Showcase page although in actual usage there would only ever be one per page:
> 
> ```
> Error: The page should have no accessibility violations. Violations:
>             [moderate]: Ensures landmarks are unique 
>             Violated 1 time. Offending nodes are: 
>             <nav class="hds-app-side-nav__list-wrapper" aria-labelledby="hds-app-side-nav-header" aria-label="Primary on portal demo 1">
> ```
> 
> I suppose I could generate a unique id instead of hard-coding as I've done currently but I would only be doing it for the Showcase instead of a real use case so that seems like overkill. Let me know what you think though. @alex-ju

Indeed, setting a unique ID will not have any benefits for our consumers.

The only way around it would be to disable the check for this specific element in our automated tests:

```
'landmark-unique': {
  enabled: false,
  selectors: [['.hds-app-side-nav__list-wrapper']],
}
```

File: packages/components/src/components/hds/app-side-nav/list/index.hbs
Comment (MelSumner) [inline packages/components/src/components/hds/app-side-nav/list/index.hbs:8]: This is where you'd add an aria-labelledby attribute, something like:

```suggestion
  <ul class="hds-app-side-nav__list" role="list" aria-labelledby="listTitle-id">
```

And then in the `app-side-nav/list-title`, add a `guidFor` that produces a unique `id` for the title. 

If we associate these correctly, the user would not have to change anything on their end.

Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/list/index.hbs]: Thanks I pushed up changes to add an id to each Title and generate a list of TitleIds that gets inserted in the aria-labelledby of the ul.

Let me know if I should remove the h3s I added earlier for the Titles.

File: packages/components/src/components/hds/app-side-nav/list/title.hbs
Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs:12]: I pushed up some changes to add ids to Titles and associate them with the `ul`.

File: packages/components/src/components/hds/app-side-nav/list/index.hbs
Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/list/index.hbs]: @MelSumner @shleewhite When you have time could you please review the latest changes I made from @MelSumner's feedback above?

File: packages/components/src/components/hds/app-side-nav/list/title.hbs
Comment (shleewhite) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs:12]: After reviewing it again, I think we should just take the work for labelling the lists by the titles out of scope for now. The reason being, now there is still one large list called "Services Default Org". 

We need to come back to this and restructure the components slightly as demonstrated in this spike: https://github.com/hashicorp/design-system/pull/2478/files

Unfortunately, I don't think there are any small wins possible for the sublists/titles without changing the structure.

I can approve this PR once it is scoped down to just adding the visually hidden H2 and labelling the nav by the H2.

Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs:12]: Thanks @shleewhite 

@MelSumner could you please let me know your input? Are you ok with what Lee said above? I'd like to know any specific changes I should make to this current PR so I can get it merged into the component branch (if you would like changes other than what Lee recommended).

Thanks!

Comment (KristinLBradley) [reply packages/components/src/components/hds/app-side-nav/list/title.hbs]: For now, I changed the `h3` I had added wrapping the Title content back into a `div`

**Review by MelSumner — APPROVED** [review]
This approach seems to work, I think we could ship this for now but also discuss Lee's sideNav spike and compare to see if we get any performance wins one way or the other. Thanks for taking the time to be thoughtful about the accessibility for this component!

**Comment by KristinLBradley** [comment]
Going to merge this into the component branch as @MelSumner approved. (There will still be time to make last minor tweaks before the component branch is merged if there's any additional feedback.)

**Pull request merged**
